csharp/src/Microsoft.ML.OnnxRuntime/OrtFloat16.shared.cs

```diff
@@ -0,0 +1,1372 @@
+            return (absValue < PositiveInfinityBits)    // is finite
+                && (absValue != 0)                      // is not zero
+                && ((absValue & BiasedExponentMask) != 0);    // is not subnormal (has a non-zero exponent)
+        }
+
+        /// <summary>
+        /// Determines whether the specified value is positive infinity.
+        /// </summary>
+        /// <param name="value">BFloat16 instance</param>
+        /// <returns></returns>
+        public static bool IsPositiveInfinity(BFloat16 value)
+        {
+            return value.value == PositiveInfinityBits;
+        }
+
+        /// <summary>
+        /// Determines whether the specified value is subnormal.
+        /// </summary>
+        /// <param name="value">BFloat16 instance</param>
+        /// <returns>true if the value is subnormal</returns>
+        public static bool IsSubnormal(BFloat16 value)
+        {
+            uint absValue = StripSign(value);
+            return (absValue < PositiveInfinityBits)    // is finite
+                && (absValue != 0)                      // is not zero
+                && ((absValue & BiasedExponentMask) == 0);    // is subnormal (has a zero exponent)
+        }
+
+        /// <summary>
+        /// Compares this object to another object, returning an integer that indicates the relationship.
+        /// </summary>
+        /// 
+        /// <param name="obj">Object to compare to</param>
+        /// <returns>A value less than zero if this is less than <paramref name="obj"/>,
+        /// zero if this is equal to <paramref name="obj"/>, or a value greater than zero
+        /// if this is greater than <paramref name="obj"/>.
+        /// </returns>
+        /// <exception cref="ArgumentException">Thrown when <paramref name="obj"/> is not of type <see cref="BFloat16"/>.</exception>
+        public int CompareTo(object obj)
+        {
+            if (!(obj is BFloat16))
+            {
+                return (obj is null) ? 1 : throw new ArgumentException("Object must be of type BFloat16");
+            }
+            return CompareTo((BFloat16)(obj));
+        }
+
+        /// <summary>
+        /// Compares this object to another object, returning an integer that indicates the relationship.
+        /// </summary>
+        /// <param name="other">Object to compare to</param>
+        /// <returns>A value less than zero if this is less than <paramref name="other"/>,
+        /// zero if this is equal to <paramref name="other"/>, 
+        /// or a value greater than zero if this is greater than <paramref name="other"/>.</returns>
+        public int CompareTo(BFloat16 other)
+        {
+            if (this < other)
+            {
+                return -1;
+            }
+
+            if (this > other)
+            {
+                return 1;
+            }
+
+            if (this == other)
+            {
+                return 0;
+            }
+
+            if (IsNaN(this))
```

Comment:
this == other already calls IsNaN, is there a way to call it only once?

Reply:
I do not know. It seemed good to model this function after System.Half.CompareTo()
[https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Half.cs,7895d5942d33f974](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Half.cs,7895d5942d33f974)